dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSchema.cpp

```diff
@@ -39,58 +40,90 @@ String ColumnFileSchema::toString() const
+
+        const Holder & holder = iter->second;
+        if (auto p = holder.column_file_schema.lock(); p)
         {
-            if (iter->second.expired())
-            {
-                iter = column_file_schemas.erase(iter);
-            }
-            else
-            {
-                ++iter;
-            }
+            // If when the item is evicted, but the item is still be used by some ColumnFiles,
+            // we increment this metrics, to show maybe the max_size of SharedBlockSchemas is not enough now.
+            GET_METRIC(tiflash_shared_block_schemas, type_still_used_when_evict).Increment();
         }
-        return true;
-    },
-                                     /*multi*/ false,
-                                     /*interval_ms*/ 60000);
-}
 
-SharedBlockSchemas::~SharedBlockSchemas()
-{
-    if (handle)
+        --current_size;
+        lru_queue.pop_front();
+        column_file_schemas.erase(iter);
+        GET_METRIC(tiflash_shared_block_schemas, type_current_size).Decrement();
+    }
+
+    if (current_size > (1ull << 63))
     {
-        background_pool.removeTask(handle);
+        throw Exception(String(__FUNCTION__) + " inconsistent, current_size < 0", ErrorCodes::LOGICAL_ERROR);
     }
 }
 
+
 ColumnFileSchemaPtr SharedBlockSchemas::find(const Digest & digest)
 {
-    std::lock_guard<std::mutex> lock(mutex);
-    auto it = column_file_schemas.find(digest);
-    if (it == column_file_schemas.end())
+    std::lock_guard lock(mutex);
+    auto iter = column_file_schemas.find(digest);
+    if (iter == column_file_schemas.end())
+    {
         return nullptr;
-    return it->second.lock();
+    }
+
+    lru_queue.splice(lru_queue.end(), lru_queue, iter->second.queue_it);
+
+    return iter->second.column_file_schema.lock();
 }
 
 ColumnFileSchemaPtr SharedBlockSchemas::getOrCreate(const Block & block)
 {
     Digest digest = hashSchema(block);
-    std::lock_guard<std::mutex> lock(mutex);
-    auto it = column_file_schemas.find(digest);
-    if (it == column_file_schemas.end() || it->second.expired())
+
+    std::lock_guard lock(mutex);
+    auto iter = column_file_schemas.find(digest);
+    if (iter != column_file_schemas.end())
     {
-        auto schema = std::make_shared<ColumnFileSchema>(block);
-        column_file_schemas.emplace(digest, schema);
-        return schema;
+        if (auto schema = iter->second.column_file_schema.lock(); schema)
+        {
+            GET_METRIC(tiflash_shared_block_schemas, type_hit_count).Increment();
+            lru_queue.splice(lru_queue.end(), lru_queue, iter->second.queue_it);
+            return schema;
+        }
+        else
+        {
+            // if the weak_ptr.lock() only get nullptr, that means the schema is not used by any ColumnFiles
+            // So we need update the item
+            GET_METRIC(tiflash_shared_block_schemas, type_miss_count).Increment();
+            std::shared_ptr<ColumnFileSchema> new_schema(new ColumnFileSchema(block));
```

Comment:
prefer using make_shared

Reply:
We don't use make_shared here, because make_shared will allocate the memory of *ptr and control block together.
Thus, the memory occupied by ColumnFileSchema will persists until all weak_ptr owners get destroyed as well.
https://lanzkron.wordpress.com/2012/04/22/make_shared-almost-a-silver-bullet/